### PR TITLE
Create package.el compatible ESS

### PR DESCRIPTION
**Pull request by jabranham**
This isn't quite done yet - I'd like to ~add tests so that it works on travis~ [DONE] and it needs documentation but figured I'd get feedback before going further.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -69,27 +69,57 @@ essr: VERSION
...
 	test -f $(ESSDIR).zip && rm -rf $(ESSDIR).zip || true
 	zip -r $(ESSDIR).zip $(ESSDIR)
 	@echo "Signing zip file"
 	$(GPG) -ba -o $(ESSDIR).zip.sig $(ESSDIR).zip
-	touch $@
+
+.PHONY: package
+package: ess-$(ESSVERSION).tar
+ess-$(ESSVERSION).tar: ess-pkg.el
+	@echo "Creating $@"
```
Comment (vspinu) [inline Makefile]: There is a new ./targets folder for make utilities. I would prefer that we move this into bash script there; it's much nicer to develop. Same for the following elisp. I think most of this stuff can be done at elisp level actually, but fine either way I guess.

File: Makefile [hunk 2]
```diff
@@ -69,27 +69,57 @@ essr: VERSION
...
+  (insert (pp-to-string
+`(define-package "ess" "$(ESSVERSION)" ; `
+   "Emacs Speaks Statistics."
+   '((emacs "25.1") ; '
+    )))))
+endef
+export package_requires
+
+ess-pkg.el:
+	@$(EMACSBATCH) --eval "$$package_requires"
```
[the remark this answers is not in the document]
Comment (vspinu) [reply Makefile]: If we can avoid build tools with a few lines of bash+elisp then it's entirely worthwhile. It's just a function call with 4 arguments after all.

**Comment by lionel-** [comment]
> I don't see a great benefit to a package.el-compatible tarball.

It standardises the packaging process in the same way that `R CMD build` provides standardised tarballs that third party packagers can rely on. I think the recent discussions in this repo have demonstrated that we need a standardised process so we don't have to interact with packagers or worry about packaging ever again.

File: VERSION [hunk 3]
```diff
@@ -1 +1 @@
-18.10.2-git
+18.10.2.999
```
[the remark this answers is not in the document]
Comment (vspinu) [reply VERSION]: It actually is. This part is controlled by `version-regexp-alist`. While snapshot is very common I think it does not resonate with the R community that much. This is why I changed `snapshot` to git. Doesn't matter much. 

I am not sure why @jabranham changed it to 999 though. If we goo with 9xx style (which could also be) we should start with .900 to allow for intermediate increments between the release cycles.

[the remark this answers is not in the document]
Comment (lionel-) [reply VERSION]: oops my bad

**Comment by vspinu** [comment]
And, as iterated in other threads, `M-x package-install-file` on that tar is actually a big deal. Then we could add a target `install-pkg` which would build the tar and install into local ELPA dir providing a cleaner alternative to #747.

File: VERSION (hunk 3 above)
[the remark this answers is not in the document]
Comment (jabranham) [reply VERSION]: Sorry, I probably should have noted this. I ran into a problem with -git and -snapshot. I spent almost no time trying to solve it so perhaps there's an easy solution. The issue I ran into is that with -git, you create `ess-$VERSION-git.tar`, but M-x package-install-file complains that it doesn't untar into `ess-$VERSIONsnapshot.tar`. Changing it to -snapshot results in the same error (note the lack of the hyphen). I can look into whether this is a bug in package.el's behavior or maybe I'm just misunderstanding something.

**Comment by dickmao** [comment]
> I think the recent discussions in this repo have demonstrated that we need a standardised process so we don't have to interact with packagers or worry about packaging ever again.

My overwhelming sense from the discussion is that our downstream colleagues will argue for a state requiring the least number of changes to their existing process.  There are enough path differences between a flat `package.el` archive and a layered debian or RPM archive that this initiative will be met with great resistance.  I hope I'm wrong.

**Comment by lionel-** [comment]
> My overwhelming sense from the discussion is that our downstream colleagues will argue for a state requiring the least number of changes to their existing process. There are enough path differences between a flat package.el archive and a layered debian or RPM archive that this initiative will be met with great resistance. 

We welcome and value the opinion of downstream packagers but they don't have the last say on this. It's true that package.el archives are less structured than R packages, but it doesn't have to be flat. We can still have our documentation or R files in subfolders, and at least we'll now exactly where they are, in `(file-name-directory load-file-name)`. This will make things much simpler.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (dickmao) [reply Makefile]: You'd need to get a lot of the metadata in `ess.el` into `ess.pkg.el` including the `julia-mode` dependency.  It's harder than you often make things sound.

**Comment by vspinu** [comment]
I am pretty sure the downstream fellows can deal with a few directory changes. It's likely that the new structure will simplify their job because it's how other elisp packages are organized. They will simply have to move the entire folder to the lisp location and optionally deal with info and doc dirs.

**Comment by lionel-** [comment]
Right. It's not about "purity", but about simplicity.

**Comment by jabranham** [comment]
To be clear - this PR doesn't change anything in the existing tarballs. It just adds a new one.

I do think we should remove the `$(ESSDIR)` target and just do something like `git archive --output=ess$(ESSVERSION).tgz` for those. It would at least make what we distribute match what github automatically generates when we do a new git tag. But, again, this PR doesn't touch the existing tarballs.

**Comment by lionel-** [comment]
yup but in the long term it'd be nice to remove all the complexity in our build infrastructure, if the package.el path proves successful.

File: VERSION (hunk 3 above)
[the remark this answers is not in the document]
Comment (jabranham) [reply VERSION]: It works just fine if you don't have a hyphen in VERSION, so I've pushed that change.

File: Makefile (hunk 1 above)
Comment (vspinu) [reply Makefile]: Yeh, those vars are problematic. Let's leave the tar as it is. 

The pkg.el creation should be in a proper lisp script though. It should take ees.el and parse metadata from there. Functionality from packae.el could be re-used but I think it's straightforward either way.

Comment (jabranham) [reply Makefile]: OK, done. There's a new script in targets/

**Comment by jabranham** [comment]
@vspinu @lionel- this branch is to the point where it produces a tar file that you can install with `package-install-file`. It also automatically tests that production and installation on our CI service (yay!). 

One thing I don't like is that it makes the structure of our development environment differ from what users will use. We'll have the lisp files in `lisp/`, while for them it'll be top-level. Perhaps this isn't a big deal (it's how MELPA functions right now IIRC), but I do wonder if we should either A) move the lisp files top-level or B) somehow set it up so that lisp/ and lisp/obsolete get added to `load-path` (perhaps with subdirs.el, though I haven't looked into if that can handle generating autoloads). WDYT?

**Comment by lionel-** [comment]
> this branch is to the point where it produces a tar file that you can install with package-install-file. It also automatically tests that production and installation on our CI service (yay!).

Nice!

> One thing I don't like is that it makes the structure of our development environment differ from what users will use. We'll have the lisp files in lisp/, while for them it'll be top-level.

Yes, I've been thinking we should eventually move the lisp files to the top-level to follow the Emacs package structure.

**Comment by vspinu** [comment]
Perfect! Looks clean and tidy. 

So we are back to `snapshot` without `-` in it? My tiny preference would be to keep the delimiter.

**Comment by jabranham** [comment]
OK, will merge this later today. I'm going to hold off on adding documentation for now since it'll have to be changed following #780 anyway. I'll open an issue and add the 19.04 milestone so we don't forget before release.

@vspinu Yes, `snapshot` without `-`. This works around a quirk of package.el/Emacs. Without getting into the weeds too much, if you go with `-snapshot`, it creates a file `ess-18.10.3-snapshot.tar`. When you go to `package-install-file`, due to the way it deconstructs the version string into a list of numbers ( see `version-regexp-alist`), it expects it to untar into `ess-18.10.3snapshot` and since it doesn't, it fails (same thing happens if you use `git` or `-git`). We could likely work around this but unless the hyphen really bothers you, it seems like it'd be more trouble than its worth.

**Pull request merged**